calicoctl/commands/ipam/show.go

```diff
@@ -19,85 +19,145 @@ import (
+	"github.com/projectcalico/libcalico-go/lib/clientv3"
 	cerrors "github.com/projectcalico/libcalico-go/lib/errors"
 	"github.com/projectcalico/libcalico-go/lib/ipam"
 
 	docopt "github.com/docopt/docopt-go"
 
 	"github.com/projectcalico/calicoctl/calicoctl/commands/argutils"
 	"github.com/projectcalico/calicoctl/calicoctl/commands/clientmgr"
+	bapi "github.com/projectcalico/libcalico-go/lib/backend/api"
 )
 
-// IPAM takes keyword with an IP address then calls the subcommands.
-func Show(args []string) error {
-	doc := constants.DatastoreIntro + `Usage:
-  calicoctl ipam show [--ip=<IP> | --show-blocks] [--config=<CONFIG>]
+type borrowedIP struct {
+	addr           string
+	borrowingNode  string
+	block          string
+	blockOwner     string
+	allocationType string
+	allocatedTo    string
+}
 
-Options:
-  -h --help             Show this screen.
-     --ip=<IP>          Report whether this specific IP address is in use.
-     --show-blocks      Show detailed information for IP blocks as well as pools.
-  -c --config=<CONFIG>  Path to the file containing connection configuration in
-                        YAML or JSON format.
-                        [default: ` + constants.DefaultConfigPath + `]
+func getBorrowedIPs(ctx context.Context, ippoolClient clientv3.IPPoolInterface, bc bapi.Client) ([]*borrowedIP, error) {
+	var details []*borrowedIP
 
-Description:
-  The ipam show command prints information about a given IP address, or about
-  overall IP usage.
-`
-	parsedArgs, err := docopt.Parse(doc, args, true, "", false, false)
+	// Read details for all blocks.
+	blocks, err := bc.List(ctx, model.BlockListOptions{}, "")
 	if err != nil {
-		return fmt.Errorf("Invalid option: 'calicoctl %s'. Use flag '--help' to read about a specific subcommand.", strings.Join(args, " "))
-	}
-	if len(parsedArgs) == 0 {
-		return nil
+		return nil, err
 	}
+	for _, kvp := range blocks.KVPairs {
+		b := kvp.Value.(*model.AllocationBlock)
 
-	ctx := context.Background()
+		for i := range b.Allocations {
+			if b.Allocations[i] != nil {
+				// Allocations[i] if not nil:
+				// - b.OrdinalToIP(i) is the allocated IP address
+				// - *b.Allocations[i] is the index of the corresponding b.Attributes
 
-	// Create a new backend client from env vars.
-	cf := parsedArgs["--config"].(string)
-	client, err := clientmgr.NewClient(cf)
+				attributes := b.Attributes[*b.Allocations[i]]
+
+				// Include both following cases:
+				// - Affinity defined and IP assigned by a different node
+				// - IP allocated from a block with no Affinity
+
+				blockOwner := ""
+				if b.Affinity != nil {
+					// Affinity is in the form host:mgianluc-bz-09s4-kadm-node-2
+					// Remove "host:"
+					parts := strings.Split(*b.Affinity, ":")
+					if len(parts) == 2 {
+						blockOwner = parts[1]
+					} else {
+						blockOwner = parts[0]
+					}
+				}
+
+				if borrowingNode, ok := attributes.AttrSecondary[model.IPAMBlockAttributeNode]; ok {
```

Comment:
One thought - for _really_ old IP allocations, we used to not set the `AttrSecondary[model.IPAMBlockAttributeNode]` at all.

That means it's possible there are IPs in the data store that don't have this set, and for those addresses we have no way to tell if they are borrowed or not. 

I think it's probably OK to do what you have in this PR - don't show them at all. 

However, we might want to consider notifying the user in some way that we detected addresses that we can't make a decision on? WDYT? I'm not sure how we'd do that.

Reply:
Maybe a message at the end would be good enough for now.

e.g. "Note: found N IP allocations without an explicit node association. Unable to detect if they are borrowed"